### PR TITLE
fix confusion matrix bug with different factor levels for response and truth

### LINKED ISSUE #2030: Potential problem in calculateConfusionMatrix()
First I'm going to apologize, as I'm not sure how to create a minimum reproducible example for this issue. I absolutely own that it may be something that I've done wrong in my (lengthy) analysis, but maybe this can just be on your radar to look out for?

The general problem is that `calculateConfusionMatrix()` is getting confused over factor levels in the result.

When I run a basic table on my data, I get the result below. The true values run down the columns and note that in my 100-sample test set, I did not have any observations of class 1 or 4 (they do exist in my training data).

```r
table(xgb_result$data$truth, xgb_result$data$response)
```
screen shot 2017-10-09 at 4 22 59 pm

When I run `calculateConfusionMatrix()`, I get the result below. Note that instead of having classes 1 and 4 missing, it is 9 and 10. The rest of the table is also messed up. (Note that in the above table, there were 11 correct classifications of class 5. Now that 11 is associated with a true value of 3.)
```r
calculateConfusionMatrix(xgb_result)
```
screen shot 2017-10-09 at 4 24 17 pm

To make sure I wasn't crazy, I wrote the table out to CSV and simply looked at the truth/response columns. The estimated model is actually doing a good job and gets the 5th class right 11 times (as shown by the `table()` method).

It's a sad consolation prize compared to a reproducible example, but here is my session info as requested:
```
R version 3.4.0 (2017-04-21)
Platform: x86_64-apple-darwin16.5.0 (64-bit)
Running under: macOS Sierra 10.12.6

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libLAPACK.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] bindrcpp_0.2       censusr_0.0.3      httr_1.2.1        
 [4] leaflet_1.1.0      RColorBrewer_1.1-2 sf_0.5-1          
 [7] parallelMap_1.3    plotly_4.7.0       kableExtra_0.4.0  
[10] knitr_1.17         mlr_2.12           ParamHelpers_1.10 
[13] MASS_7.3-47        dplyr_0.7.3        purrr_0.2.3       
[16] readr_1.1.0        tidyr_0.7.1        tibble_1.3.4      
[19] ggplot2_2.2.1      tidyverse_1.1.1    pixiedust_0.8.1   
[22] h2o_3.10.5.3      

loaded via a namespace (and not attached):
 [1] nlme_3.1-131        bitops_1.0-6        lazyWeave_3.0.1    
 [4] lubridate_1.6.0     rprojroot_1.2       tools_3.4.0        
 [7] backports_1.1.0     R6_2.2.2            rpart_4.1-11       
[10] Hmisc_4.0-3         DBI_0.6-1           lazyeval_0.2.0     
[13] colorspace_1.3-2    nnet_7.3-12         tidyselect_0.2.0   
[16] gridExtra_2.2.1     mnormt_1.5-5        compiler_3.4.0     
[19] rvest_0.3.2         htmlTable_1.9       xml2_1.1.1         
[22] scales_0.4.1        checkmate_1.8.3     psych_1.7.3.21     
[25] stringr_1.2.0       digest_0.6.12       foreign_0.8-67     
[28] rmarkdown_1.6       base64enc_0.1-3     pkgconfig_2.0.1    
[31] htmltools_0.3.6     htmlwidgets_0.8     rlang_0.1.2        
[34] readxl_1.0.0        BBmisc_1.11         shiny_1.0.3        
[37] bindr_0.1           jsonlite_1.4        crosstalk_1.0.0    
[40] acepack_1.4.1       RCurl_1.95-4.8      magrittr_1.5       
[43] Formula_1.2-1       Matrix_1.2-9        Rcpp_0.12.12       
[46] munsell_0.4.3       stringi_1.1.5       yaml_2.1.14        
[49] plyr_1.8.4          grid_3.4.0          parallel_3.4.0     
[52] forcats_0.2.0       udunits2_0.13       lattice_0.20-35    
[55] haven_1.0.0         splines_3.4.0       hms_0.3            
[58] xgboost_0.6-4       reshape2_1.4.2      glue_1.1.1         
[61] evaluate_0.10.1     latticeExtra_0.6-28 data.table_1.10.4  
[64] modelr_0.1.0        httpuv_1.3.3        cellranger_1.1.0   
[67] gtable_0.2.0        assertthat_0.2.0    mime_0.5           
[70] xtable_1.8-2        broom_0.4.2         survival_2.41-3    
[73] viridisLite_0.2.0   units_0.4-6         cluster_2.0.6 
```

### PR DESCRIPTION
**Pull request by larskotthoff**
Fixes #2030

### CODE REVIEW

**Comment by berndbischl** [comment]
i am unsure whether that is the correct point to fix it. shouldnt we ensure "up highter" that the levels on the pred object are correct?

**Comment by berndbischl** [comment]
pls take a look at Prediction.R

```
  # truth can come from a simple "newdata" df. then there might not be all factor levels present
  if (!is.null(truth))
    levels(truth) = union(levels(truth), task.desc$class.levels)
```

it seems to be that we also need to ensure the ORDER of the levels there?

**Comment by larskotthoff** [comment]
I was wondering about this as well. The confusion matrix seems to be the only thing that's affected by it though, and changing the levels in Prediction.R didn't always work -- seems that when probabilities are predicted, the response is added somewhere else.

**Comment by ja-thomas** [comment]
Looks good, not sure if we want to fix it someplace else.

**Comment by larskotthoff** [comment]
Merging. We can always revert this later.
